Fix MAGMA qr for empty batched inputs

IvanYashchuk opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #57317 Add cuSOLVER path for torch.linalg.lstsq
* #57316 Add CUDA support for torch.ormqr
* #57315 Port CPU torch.ormqr to ATen
* #57314 Fix torch.ormqr for non Fortran-contiguous inputs
* **#56257 Fix MAGMA qr for empty batched inputs**
* #56256 Add cuSOLVER path for torch.linalg.qr
* #56255 Remove size arguments for internal orgqr and geqrf calls
* #56254 Add non-allocating helper function for torch.linalg.qr

CPU and cuSOLVER path were fixed with refactoring of
`_linalg_qr_helper_default`.

Resolves https://github.com/pytorch/pytorch/issues/50576

Differential Revision: [D27960157](https://our.internmc.facebook.com/intern/diff/D27960157)

IvanYashchuk committed:
Fix MAGMA qr for empty batched inputs

CPU and cuSOLVER path were fixed with refactoring of
`_linalg_qr_helper_default`.

Resolves https://github.com/pytorch/pytorch/issues/50576

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Fix MAGMA qr for empty batched inputs"


CPU and cuSOLVER path were fixed with refactoring of
`_linalg_qr_helper_default`.

Resolves https://github.com/pytorch/pytorch/issues/50576

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Fix MAGMA qr for empty batched inputs"


CPU and cuSOLVER path were fixed with refactoring of
`_linalg_qr_helper_default`.

Resolves https://github.com/pytorch/pytorch/issues/50576

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Fix MAGMA qr for empty batched inputs"


CPU and cuSOLVER path were fixed with refactoring of
`_linalg_qr_helper_default`.

Resolves https://github.com/pytorch/pytorch/issues/50576

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Fix MAGMA qr for empty batched inputs"


CPU and cuSOLVER path were fixed with refactoring of
`_linalg_qr_helper_default`.

Resolves https://github.com/pytorch/pytorch/issues/50576

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Fix MAGMA qr for empty batched inputs"


CPU and cuSOLVER path were fixed with refactoring of
`_linalg_qr_helper_default`.

Resolves https://github.com/pytorch/pytorch/issues/50576

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Fix MAGMA qr for empty batched inputs"


CPU and cuSOLVER path were fixed with refactoring of
`_linalg_qr_helper_default`.

Resolves https://github.com/pytorch/pytorch/issues/50576

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Fix MAGMA qr for empty batched inputs"


CPU and cuSOLVER path were fixed with refactoring of
`_linalg_qr_helper_default`.

Resolves https://github.com/pytorch/pytorch/issues/50576

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Fix MAGMA qr for empty batched inputs"


CPU and cuSOLVER path were fixed with refactoring of
`_linalg_qr_helper_default`.

Resolves https://github.com/pytorch/pytorch/issues/50576

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Fix MAGMA qr for empty batched inputs"


CPU and cuSOLVER path were fixed with refactoring of
`_linalg_qr_helper_default`.

Resolves https://github.com/pytorch/pytorch/issues/50576

[ghstack-poisoned]

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -2170,8 +2170,8 @@ def sample_inputs_linalg_qr(op_info, device, dtype, requires_grad=False, **kwarg
     # TODO: add 0 to 'ns' and (0, ) to 'batches'
```
Does this fix address the cited issue? https://github.com/pytorch/pytorch/issues/50576

IvanYashchuk replied:
It does. I forgot to remove this TODO.

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -2170,8 +2170,8 @@ def sample_inputs_linalg_qr(op_info, device, dtype, requires_grad=False, **kwarg
     # https://github.com/pytorch/pytorch/issues/50576
-    batches = [(), (2, ), (1, 1)]
-    ns = [2, 5]
+    batches = [(), (0,), (2, ), (1, 1)]
```
Updating sample inputs is cool, but I think we should extend the test_linalg.py tests for this, too?

IvanYashchuk replied:
You're right, I'll update it.

IvanYashchuk committed:
Update on "Fix MAGMA qr for empty batched inputs"


CPU and cuSOLVER path were fixed with refactoring of
`_linalg_qr_helper_default`.

Resolves https://github.com/pytorch/pytorch/issues/50576

Differential Revision: [D27960157](https://our.internmc.facebook.com/intern/diff/D27960157)

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Fix MAGMA qr for empty batched inputs"


CPU and cuSOLVER path were fixed with refactoring of
`_linalg_qr_helper_default`.

Resolves https://github.com/pytorch/pytorch/issues/50576

Differential Revision: [D27960157](https://our.internmc.facebook.com/intern/diff/D27960157)

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Fix MAGMA qr for empty batched inputs"


CPU and cuSOLVER path were fixed with refactoring of
`_linalg_qr_helper_default`.

Resolves https://github.com/pytorch/pytorch/issues/50576

Differential Revision: [D27960157](https://our.internmc.facebook.com/intern/diff/D27960157)

[ghstack-poisoned]

IvanYashchuk commented:
@mruberry, I extended the `test_linalg.py` tests with single matrix empty and batched empty inputs. The tests are failing on master and this PR resolves the failures.

mruberry reviewed (approved):
Cool!

I've asked @xwang233 and @Lezcano to help with the two unapproved PRs in the unmerged half of this stack.

IvanYashchuk committed:
Update on "Fix MAGMA qr for empty batched inputs"


CPU and cuSOLVER path were fixed with refactoring of
`_linalg_qr_helper_default`.

Resolves https://github.com/pytorch/pytorch/issues/50576

Differential Revision: [D27960157](https://our.internmc.facebook.com/intern/diff/D27960157)

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Fix MAGMA qr for empty batched inputs"


CPU and cuSOLVER path were fixed with refactoring of
`_linalg_qr_helper_default`.

Resolves https://github.com/pytorch/pytorch/issues/50576

Differential Revision: [D27960157](https://our.internmc.facebook.com/intern/diff/D27960157)

[ghstack-poisoned]
